### PR TITLE
Use custom scopes appropriately

### PR DESCRIPTION
**Pull request by sgyang**
The activity has to retain the reference of its component.
If you instantiate ActivityModule every time on the call of MainApplication::getComponent(), the scope will be broken; each child fragment has activity-scoped instances **separately** despite their life cycle should be tied to their activity.
Actually the custom scopes are not needed in this app and that's why there are no bugs related to them. So this PR is some kind of a reference.

### CODE REVIEW

**Comment by konifar** [comment]
Thanks for your PR! I'll check tonight :)

File: app/src/main/java/io/github/droidkaigi/confsched/activity/BaseActivity.java [hunk 1]
```diff
@@ -0,0 +1,22 @@
+package io.github.droidkaigi.confsched.activity;
+
+import android.support.annotation.NonNull;
+import android.support.v7.app.AppCompatActivity;
+
+import io.github.droidkaigi.confsched.MainApplication;
+import io.github.droidkaigi.confsched.di.ActivityComponent;
+import io.github.droidkaigi.confsched.di.ActivityModule;
+
+public class BaseActivity extends AppCompatActivity {
```
Comment (hotchemi) [inline app/src/main/java/io/github/droidkaigi/confsched/activity/BaseActivity.java]: How about make this class abstract?

File: app/src/main/java/io/github/droidkaigi/confsched/fragment/BaseFragment.java [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+import android.app.Activity;
+import android.support.annotation.NonNull;
+import android.support.v4.app.Fragment;
+
+import io.github.droidkaigi.confsched.activity.BaseActivity;
+import io.github.droidkaigi.confsched.di.FragmentComponent;
+import io.github.droidkaigi.confsched.di.FragmentModule;
+
+public class BaseFragment extends Fragment {
```
Comment (hotchemi) [inline app/src/main/java/io/github/droidkaigi/confsched/fragment/BaseFragment.java]: https://github.com/konifar/droidkaigi2016/pull/317/files#r53754226

**Comment by sgyang** [comment]
Yes I know this change is little bit disruptive, but I don't know how to manage their components properly except this way. In fact I'm not sure if it is a correct manner.
Any advice is welcome!

File: app/src/main/java/io/github/droidkaigi/confsched/activity/BaseActivity.java [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+import io.github.droidkaigi.confsched.di.ActivityComponent;
+import io.github.droidkaigi.confsched.di.ActivityModule;
+
+public abstract class BaseActivity extends AppCompatActivity {
+
+    private ActivityComponent activityComponent;
+
+    @NonNull
+    public ActivityComponent getComponent() {
+        if (activityComponent == null) {
```
Comment (konifar) [inline app/src/main/java/io/github/droidkaigi/confsched/activity/BaseActivity.java]: Sorry, I could not understand an intention of your change :bow: 
When does this `activityComponent` become not null? It seems to be called in `onCreate()` method one time.

Comment (sgyang) [inline app/src/main/java/io/github/droidkaigi/confsched/activity/BaseActivity.java]: Attached fragments will call this method on their `getComponent()` (and so it is public).
`activityComponent` should be a *singleton in the activity scope*. So I created the base classes and added this methods.

Comment (konifar) [inline app/src/main/java/io/github/droidkaigi/confsched/activity/BaseActivity.java]: I understood. Thanks

**Comment by konifar** [comment]
Looks good! But I'm not sure that this is best way...
So I'll merge it, and wait anyone who know better way comes.

Anyway thanks for your fix :+1:

**Comment by sgyang** [comment]
Yes! In fact I'm not sure too. :joy_cat: 
As another example, [Fernando Cejas's Clean architecture](https://github.com/android10/Android-CleanArchitecture) introduces both `BaseActivity` and `BaseFragment`, but they are not used for retaining scoped instances. In this case, `HasComponent<C>` interface plays the role. Each activity implements `HasComponent<C>` and then attached fragments obtain its parent component instance through this interface. This way may be a more general approach, but it seems complicated.
Thanks! :+1:

**Pull request merged**